tests/test_qutrit_device.py

```diff
@@ -0,0 +1,670 @@
+        m.setattr(QutritDevice, "short_name", "MockQutritDevice")
+        m.setattr(QutritDevice, "operations", ["QutritUnitary", "Identity"])
+        m.setattr(QutritDevice, "observables", ["Identity"])
+        m.setattr(QutritDevice, "expval", lambda self, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "var", lambda self, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "sample", lambda self, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "state", 0)
+        m.setattr(QutritDevice, "density_matrix", lambda self, wires=None: 0)
+        m.setattr(QutritDevice, "probability", lambda self, wires=None, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "apply", lambda self, x, **kwargs: x)
+
+        def get_qutrit_device(wires=1):
+            return QutritDevice(wires=wires)
+
+        yield get_qutrit_device
+
+
+@pytest.fixture(scope="function")
+def mock_qutrit_device_shots(monkeypatch):
+    """A function to create a mock device that mocks the methods related to
+    statistics (expval, var, sample, probability)"""
+    with monkeypatch.context() as m:
+        m.setattr(QutritDevice, "__abstractmethods__", frozenset())
+        m.setattr(QutritDevice, "_capabilities", mock_qutrit_device_capabilities)
+        m.setattr(QutritDevice, "short_name", "MockQutritDevice")
+        m.setattr(QutritDevice, "operations", ["QutritUnitary", "Identity"])
+        m.setattr(QutritDevice, "observables", ["Identity"])
+        m.setattr(QutritDevice, "expval", lambda self, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "var", lambda self, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "sample", lambda self, *args, **kwargs: 0)
+        m.setattr(QutritDevice, "state", 0)
+        m.setattr(QutritDevice, "density_matrix", lambda self, wires=None: 0)
+        m.setattr(QutritDevice, "apply", lambda self, x, **kwargs: x)
+        m.setattr(
+            QutritDevice,
+            "analytic_probability",
+            lambda self, wires=None: QutritDevice.marginal_prob(self, np.ones(9) / 9.0, wires),
+        )
+
+        def get_qutrit_device(wires=1, shots=None):
+            return QutritDevice(wires=wires, shots=shots)
+
+        yield get_qutrit_device
+
+
+@pytest.fixture(scope="function")
+def mock_qutrit_device_with_original_statistics(monkeypatch):
+    """A function to create a mock qutrit device that uses the original statistics related
+    methods"""
+    with monkeypatch.context() as m:
+        m.setattr(QutritDevice, "__abstractmethods__", frozenset())
+        m.setattr(QutritDevice, "_capabilities", mock_qutrit_device_capabilities)
+        m.setattr(QutritDevice, "short_name", "MockQutritDevice")
+        m.setattr(QutritDevice, "operations", ["QutritUnitary", "Identity"])
+        m.setattr(QutritDevice, "observables", ["Identity"])
+
+        def get_qutrit_device(wires=1):
+            return QutritDevice(wires=wires)
+
+        yield get_qutrit_device
+
```

Comment:
Can you remind me why the equivalent of `mock_qubit_device_with_paulis_and_methods` and `mock_qubit_device_with_paulis_rotations_and_methods` are missing? Are those for testing features that are not implemented yet? In case yes, you could already add the tests and mark them with `@pytest.mark.xfail`

Reply:
As the only currently supported qutrit operation is `qml.QutritUnitary`, such mock devices can't be used yet. However, upon closer inspection of how `test_qubit_device.py` actually uses `mock_qubit_device_with_paulis_and_methods` and `mock_qubit_device_with_paulis_rotations_and_methods`, I believe that they won't be very necessary in the future either, as they are largely used to ensure that the execution flow is as expected rather than actually checking the correctness of the operations themselves (which makes sense; `QubitDevice` and `QutritDevice` are abstract classes and there is no internal state that operations are applied to), and this type of checking can be done using `QutritUnitary`, or any other arbitrary operations (if they were available) for that matter.

Reply:
A small flyby comment: the usage of the `mock_qubit_device_with_*` pattern is quite outdated in the test suite, and we've long wanted to replace this with a _single_ mock device with all operations mocked out. We just haven't gotten around to it yet :)